source/common/config/config_provider_impl.h

```diff
@@ -184,25 +190,45 @@ class ConfigSubscriptionCommonBase : protected Logger::Loggable<Logger::Id::conf
     init_target_.ready();
   }
 
-  /**
-   * Returns one of the bound mutable config providers.
-   * @return const MutableConfigProviderCommonBase* a const pointer to a
-   *         bound MutableConfigProviderCommonBase or nullptr when there are none.
-   */
-  MutableConfigProviderCommonBase* getAnyBoundMutableConfigProvider() const {
-    return !mutable_config_providers_.empty() ? *mutable_config_providers_.begin() : nullptr;
-  }
-
 protected:
+  struct ThreadLocalConfig : public ThreadLocal::ThreadLocalObject {
+    ThreadLocalConfig(ConfigProvider::ConfigConstSharedPtr initial_config)
+        : config_(std::move(initial_config)) {}
+
+    ConfigProvider::ConfigConstSharedPtr config_;
+  };
+
   ConfigSubscriptionCommonBase(const std::string& name, const uint64_t manager_identifier,
                                ConfigProviderManagerImplBase& config_provider_manager,
-                               TimeSource& time_source, const SystemTime& last_updated,
-                               const LocalInfo::LocalInfo& local_info)
-      : name_(name),
+                               Server::Configuration::FactoryContext& factory_context)
+      : name_(name), tls_(factory_context.threadLocal().allocateSlot()),
         init_target_(fmt::format("ConfigSubscriptionCommonBase {}", name_), [this]() { start(); }),
         manager_identifier_(manager_identifier), config_provider_manager_(config_provider_manager),
-        time_source_(time_source), last_updated_(last_updated) {
-    Envoy::Config::Utility::checkLocalInfo(name, local_info);
+        time_source_(factory_context.timeSource()),
+        last_updated_(factory_context.timeSource().systemTime()) {
+    Envoy::Config::Utility::checkLocalInfo(name, factory_context.localInfo());
+  }
+
+  /**
+   * Propagates a config update to worker threads.
+   *
+   * @param update_fn the callback to run on each thread, it takes the previous version Config and
+   * returns a updated/new version Config.
+   * @param complete_cb the callback to run when the update propagation is done.
+   */
+  void applyConfigUpdate(
```

Comment:
I like the simplification introduced by generalizing this function. It avoids delta vs non-delta code bloat which was a pending TODO.

With that said, after thinking about this quite a bit, I think the "ConfigImpl" and the thread local storage on which it is stored are a better fit to be owned by the `MutableConfigProviderCommonBase`. This maintains the existing boundary between the \*ConfigSubscription* base classes which manage the underlying xDS subscription, and the \*ConfigProvider* base classes which manage the `ConfigConstSharedPtr` resulting from the config update. As it stands now with these changes, the \*ConfigProvider* classes are just a thin shim to provide access to the ConfigSubscription and could simply be removed.

The proposal above would reintroduce the notion of bound config providers to a subscription, but I propose a new simplification here: instead of having multiple config providers bound to a subscription, have a _single_ bound provider which owns the `tls_` and associated thread local `ConfigConstSharedPtr`s. Instead of `ConfigProviderManager::createXdsConfigProvider` returning a new instance of a `ConfigProviderPtr` on every call, it would instead return a reference to the instance bound to the subscription (allocating a new instance if a bound one is not found). 

Effectively, this would translate into moving `applyConfigUpdate` to `MutableConfigProviderCommonBase` along with `tls_`. `MutableConfigProviderBase` and `DeltaMutableConfigProvider` are still removed in this proposal.

The modified approach would achieve several goals:

1. Maintains the boundary between \*ConfigSubscription* classes managing the xDS subscription and the \*ConfigProvider* classes managing the resulting config objects.
2. Retains most of the simplification introduced by these changes and doesn't fundamentally change the approach, allowing most of the code to simply be moved.
3. Creates an API that enforces users of the ConfigProvider framework to think about the desired lifetime of the `ConfigConstSharedPtr` returned by the `update_fn`. This is beneficial since one of the key goals of introducing this framework is to minimize proliferation of copies of the config across consumers of the config.

WDYT?

Reply:
we had an offline discussion, the way proposed in above comment has a problem around the ownership of subscription&Providers, it can't guarantee a subscription gets teared down when there are no users(providers).

We agreed that having a subscription maintains/owns the Config instance(s) make more sense for:
    - Simpler update propagation: providers no need to worry about Config impl update at all.
    - enforced x-providers Config sharing (which is the design goal of the provider framework)
    - Keeps the same ownership mode as the existing RDS does, a subscription is shared between providers to ensure its life circle align with users'.